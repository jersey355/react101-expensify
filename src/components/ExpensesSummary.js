import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export class ExpensesSummary extends React.Component {

    getTotal = (expenses) => {
        const total = this.props.getTotal(expenses);
        this.props.history.push('/');
        return total;
    }

    render() {
        const expCount = this.props.expenses ? this.props.expenses.length : 0;
        const expTotal = numeral(getTotal(this.props.expenses) / 100).format('$0,0.00');
        return (
            <div>
                <p>{`Viewing ${expCount} totalling ${expTotal}`}</p>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
});

const mapDispatchToProps = (dispatch) => ({
    getTotal: (expenses) => dispatch(getTotal(expenses))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpensesSummary);