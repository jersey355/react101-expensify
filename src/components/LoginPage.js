import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogin, startGoogleLogin, startFacebookLogin } from '../actions/auth';

export class LoginPage extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            email: props.email ? props.email : '',
            password: props.password ? props.password : '',
            error: ''
        };
    }

    onSubmit = (e) => {

        e.preventDefault();

        if (!this.state.email || !this.state.password) {
            this.setState(() => ({ error: 'Email and password are required for login!' }));
            return;
        }

        this.setState(() => ({ error: '' }));
        this.props.startLogin(this.state.email, this.state.password);
    };

    onEmailChange = (e) => {
        const email = e.target.value;
        this.setState(() => ({ email }));
    }

    onPasswordChange = (e) => {
        const password = e.target.value;
        this.setState(() => ({ password }));
    }

    render() {
        return (
            <div className="box-layout">
                <div className="box-layout__box">
                    <h1 className="box-layout__title">Expensify</h1>
                    <p>It's time to get your expenses online!</p>
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <form onSubmit={this.onSubmit}>
                        <input
                            type="text"
                            placeholder="Email Address"
                            autoFocus
                            className="text-input"
                            onChange={this.onEmailChange}
                        />
                        <input
                            type="password"
                            placeholder="Password"
                            className="text-input text-area--stacked"
                            onChange={this.onPasswordChange}
                        />
                        <div className="box-layout__dual-column-content">
                            <button className="button button--link2">Login</button>
                            <Link className="button button--link2" to="/register">Register</Link>
                        </div>
                    </form>
                    <button className="button button--stacked" onClick={this.props.startGoogleLogin}>Login with Google</button>
                    <button className="button button--stacked" onClick={this.props.startFacebookLogin}>Login with Facebook</button>
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startLogin: (email, password) => dispatch(startLogin(email, password)),
    startGoogleLogin: () => dispatch(startGoogleLogin()),
    startFacebookLogin: () => dispatch(startFacebookLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);