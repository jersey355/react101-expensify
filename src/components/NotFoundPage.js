import React from 'react';

const NotFoundPage = () => (
    <div>
        <p>404 - NOT FOUND!</p>
    </div>
);

export default NotFoundPage;