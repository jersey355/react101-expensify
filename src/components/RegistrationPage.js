import React from 'react';
import { connect } from 'react-redux';
import PublicHeader from './PublicHeader';
import { startUserRegistration } from '../actions/auth';

export class RegistrationPage extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            email: props.email ? props.email : '',
            password1: props.password1 ? props.password1 : '',
            password2: props.password2 ? props.password2 : '',
            error: ''
        };
    }

    onEmailChange = (e) => {
        const email = e.target.value;
        this.setState(() => ({ email }));
    }

    onPassword1Change = (e) => {
        const password1 = e.target.value;
        this.setState(() => ({ password1 }));
    }

    onPassword2Change = (e) => {
        const password2 = e.target.value;
        this.setState(() => ({ password2 }));
    }

    onSubmit = (e) => {

        e.preventDefault();

        if (!this.state.email || !this.state.password1 || !this.state.password2) {
            this.setState(() => ({ error: 'All fields are required!' }));
            return;
        }

        if (this.state.password1 !== this.state.password2) {
            this.setState(() => ({ error: 'Passwords must match (including case)!' }));
            return;
        }

        this.setState(() => ({ error: '' }));
        this.props.startUserRegistration(this.state.email, this.state.password1);
    };

    render() {
        return (
            <div>
                <PublicHeader />
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Register</h1>
                    </div>
                </div>
                <div className="content-container">
                    <form className="form" onSubmit={this.onSubmit}>
                        <input
                            type="text"
                            placeholder="Email Address"
                            autoFocus
                            className="text-input"
                            value={this.state.email}
                            onChange={this.onEmailChange}
                        />
                        <input
                            type="password"
                            placeholder="Password"
                            className="text-input"
                            value={this.state.password1}
                            onChange={this.onPassword1Change}
                        />
                        <input
                            type="password"
                            placeholder="Confirm Password"
                            className="text-input"
                            value={this.state.password2}
                            onChange={this.onPassword2Change}
                        />
                        <div className="box-layout__dual-column-content">
                            <button className="button">Register</button>
                            <button className="button button--secondary">Cancel</button>
                        </div>
                    </form>
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startUserRegistration: (email, password) => dispatch(startUserRegistration(email, password))
});

export default connect(undefined, mapDispatchToProps)(RegistrationPage);