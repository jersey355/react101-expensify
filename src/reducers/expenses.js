const expensesDefaultState = [];

export default (state = expensesDefaultState, action) => {

    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            // note: we use map to update only the target expense, then 
            // we use the spread operator to first copy original expense, then override
            // with updates
            return state.map((exp) => {
                if (exp.id !== action.id) return exp;
                return { ...exp, ...action.updates };
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }

};
