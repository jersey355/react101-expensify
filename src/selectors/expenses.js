import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((exp) => {
        const createdAtMoment = moment(exp.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = exp.description && exp.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        } else {
            return a.createdAt < b.createdAt ? 1 : -1;
        }
    });
};