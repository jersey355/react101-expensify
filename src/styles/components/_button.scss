.button {
    background: $blue;
    border: none;
    color: white;
    display: inline-block;
    font-size: $font-size-large;
    font-weight: 300;
    line-height: 1;
    padding: $s-size;
    text-align: center;
    text-decoration: none;
}

.button--secondary {
    background: #888;
}

.button--link {
    background: none;
}

.button--link2 {
    background: none;
    color: $blue;
}

.button--stacked {
    margin-top: $s-size;
    width: 100%;
}