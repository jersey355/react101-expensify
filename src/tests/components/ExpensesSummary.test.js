import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('Should render expense summary with expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render expense summary with no expenses', () => {
    const wrapper = shallow(<ExpensesSummary />);
    expect(wrapper).toMatchSnapshot();
});