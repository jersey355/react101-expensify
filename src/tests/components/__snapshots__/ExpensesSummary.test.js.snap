// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense summary with expenses 1`] = `
<div>
  <p>
    Viewing 3 totalling $1,141.95
  </p>
</div>
`;

exports[`Should render expense summary with no expenses 1`] = `
<div>
  <p>
    Viewing 0 totalling $0.00
  </p>
</div>
`;
